transports/dns/src/lib.rs

```diff
@@ -238,6 +225,90 @@ where TErr: error::Error + 'static
     }
 }
 
+async fn resolve_dns<T>(p: Protocol<'_>, suffix: Multiaddr) -> Result<Multiaddr, DnsErr<T::Error>>
+    where
+        T: Transport + Send + 'static,
+        T::Error: Send,
+        T::Dial: Send
+{
+    match p {
+        Protocol::Dns4(ref n) | Protocol::Dns6(ref n) => {
+            let name = n.to_string();
+            let to_resolve = format!("{}:0", name);
+            let list = match to_resolve[..].to_socket_addrs() {
+                Ok(list) => Ok(list.map(|s| s.ip()).collect::<Vec<_>>()),
+                Err(e) => Err(e)
+            }.map_err(|_| {
+                error!("DNS resolver crashed");
+                DnsErr::ResolveFail(name.clone())
+            })?;
+
+            let is_dns4 = if let Protocol::Dns4(_) = p { true } else { false };
+
+            list.into_iter()
+                .filter_map(|addr| {
+                    if (is_dns4 && addr.is_ipv4()) || (!is_dns4 && addr.is_ipv6()) {
+                        Some(Multiaddr::from(addr))
+                    } else {
+                        None
+                    }
+                })
+                .next()
+                .ok_or_else(|| DnsErr::ResolveFail(name))
+        }
+        Protocol::Dnsaddr(ref n) => {
+            let conn = UdpClientConnection::new("8.8.8.8:53".parse().unwrap()).unwrap(); // TODO: error handling
+            let client = SyncClient::new(conn); // TODO: should use async client?
```

Comment:
> // TODO: should use async client?

Yes!

Reply:
Hmm... 

I've added [trust-dns](https://github.com/bluejekyll/trust-dns) in order to query TXT records. I'm trying to switch the SyncClient to AsyncClient but I think AsyncClient can not be used for rust-libp2p because a background future is required for the client. 🤔

https://docs.rs/trust-dns-client/0.19.3/trust_dns_client/#async-usage

```rust
// Create a new client, the bg is a background future which handles
//   the multiplexing of the DNS requests to the server.
//   the client is a handle to an unbounded queue for sending requests via the
//   background. The background must be scheduled to run before the client can
//   send any dns requests
let client = AsyncClient::connect(stream);

// await the connection to be established
let (mut client, bg) = runtime.block_on(client).expect("connection failed");

// make sure to run the background task
runtime.spawn(bg);
```

======

Do you know other library what can query TXT records asynchronously? 💦

Reply:
> I think AsyncClient can not be used for rust-libp2p because a background future is required for the client.

That makes it more difficult to implement, but normally not impossible.

Reply:
The idea is not to run the background future in the background, but to run it alongside with the "front" one. I don't have more precise guidelines without looking deeply into details (which I don't really have the time right now to do) though.

Reply:
I've thought for a while about this but I can not come up anything to address the issue... hmm... 🤔💦